borsh-rs/borsh/src/schema.rs

```diff
@@ -0,0 +1,387 @@
+impl_tuple!(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14);
+impl_tuple!(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15);
+impl_tuple!(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16);
+impl_tuple!(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17);
+impl_tuple!(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18);
+impl_tuple!(
+    T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19
+);
+impl_tuple!(
+    T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11, T12, T13, T14, T15, T16, T17, T18, T19, T20
+);
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! map(
+    () => { ::std::collections::HashMap::new() };
+    { $($key:expr => $value:expr),+ } => {
+        {
+            let mut m = ::std::collections::HashMap::new();
+            $(
+                m.insert($key.to_string(), $value.to_string());
+            )+
+            m
+        }
+     };
+    );
+
+    #[test]
+    fn simple_option() {
+        let actual_name = Option::<u64>::schema_type_name();
+        let mut actual_defs = map!();
+        Option::<u64>::add_rec_type_definitions(&mut actual_defs);
+        assert_eq!("Option<u64>", actual_name);
+        assert_eq!(
+            map! {"Option<u64>" => r#"{ "kind": "enum", "variants": [ ["None", "nil"], ["Some", "u64"] ] }"#},
+            actual_defs
+        );
+    }
+
+    #[test]
+    fn nested_option() {
+        let actual_name = Option::<Option<u64>>::schema_type_name();
+        let mut actual_defs = map!();
+        Option::<Option<u64>>::add_rec_type_definitions(&mut actual_defs);
+        assert_eq!("Option<Option<u64>>", actual_name);
+        assert_eq!(
+            map! {
+            "Option<u64>" => r#"{ "kind": "enum", "variants": [ ["None", "nil"], ["Some", "u64"] ] }"#,
+            "Option<Option<u64>>" => r#"{ "kind": "enum", "variants": [ ["None", "nil"], ["Some", "Option<u64>"] ] }"#
+            },
+            actual_defs
+        );
+    }
+
+    #[test]
+    fn simple_vec() {
+        let actual_name = Vec::<u64>::schema_type_name();
+        let mut actual_defs = map!();
+        Vec::<u64>::add_rec_type_definitions(&mut actual_defs);
+        assert_eq!("Vec<u64>", actual_name);
+        assert_eq!(
+            map! {"Vec<u64>" => r#"{ "kind": "sequence", "params": [ "u64" ] }"#},
+            actual_defs
+        );
+    }
+
+    #[test]
+    fn nested_vec() {
+        let actual_name = Vec::<Vec<u64>>::schema_type_name();
+        let mut actual_defs = map!();
+        Vec::<Vec<u64>>::add_rec_type_definitions(&mut actual_defs);
+        assert_eq!("Vec<Vec<u64>>", actual_name);
+        assert_eq!(
+            map! {
+            "Vec<u64>" => r#"{ "kind": "sequence", "params": [ "u64" ] }"#,
+            "Vec<Vec<u64>>" => r#"{ "kind": "sequence", "params": [ "Vec<u64>" ] }"#
```

Comment:
This looks inconsistent to me. Why do we encode types with `kind` (`sequence` / `tuple` / `enum`) on the top-level, and then in params we use `Vec<...>`?

Reply:
Kind does not encode the type, it is [a type constructor](https://en.wikipedia.org/wiki/Kind_(type_theory)). Meaning knowing the value of the kind and the params we know how to construct this type in memory. Arguably we could leave just four kinds: `enum` (a.k.a tagged union), `sequence`, `array`, `tuple` (and merge `struct` with `tuple`). `enum` requires reading the byte that tells how many variants are there. `sequence` requires reading the byte that tell how many elements are there. `array` does not require reading additional bytes but it has uniform type of element. `tuple` and `struct` have non-uniform types. However I decided to keep `struct` because it allows us to have explicit field names which will be useful when we will be converting from JSON to Borsh.

Reply:
Relevant, but not exactly the same: https://serde.rs/data-model.html

Reply:
I am having hard times understanding why we need this top-level type [constructor] be different from the inner items. To achieve consistency, I see two ways:

```
// 1
"Vec<Vec<u64>>" => r#"{ "kind": "sequence", "params": [ {"kind": "sequence", "params": ["u64"] } ] }"#

// 2
"Vec<Vec<u64>>" => r#"Vec<Vec<u64>>"#
```

I may just miss something, so I will approve the PR (it looks solid from the code and tests point of view), and let you make the decision. I am ready to continue the discussion, though, if you feel that there is a benefit of ELI5.

Reply:
The idea is that every entry of the map is the following: `<name of the type> => <how to construct it>`. We cannot have
```
"Vec<Vec<u64>>" => r#"{ "kind": "sequence", "params": [ {"kind": "sequence", "params": ["u64"] } ] }"#
```
and should instead have:
```
"Vec<Vec<u64>>" => r#"{ "kind": "sequence", "params": [ "Vec<u64>" ] }"#,
"Vec<u64>" => r#"{ "kind": "sequence", "params": [ "u64" ] }"#
```
because we can have recursive definitions like:
```rust
struct Node {
parent: Rc<Node>,
children: Vec<Rc<Node>>,
}
```
that cannot be described using the proposed way.

Also we cannot have 
```
"Vec<Vec<u64>>" => r#"Vec<Vec<u64>>"#
```
for the same reason, and also because processing it would require parsing the string, which is much harder than simply working with JSON.

Reply:
Another reason is that if we have `Vec<A>` and `Hashmap<A, B>` we don't want `A` to be redefined twice.

Reply:
Aaaah, I got it! `Vec<u64>` is essentially an alias, which we use inside `Vec<Vec<u64>>` definition! Now it makes perfect sense. Looks good to me!